src/lib/mdns/ServicePool.h

```diff
@@ -0,0 +1,140 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include <lib/mdns/platform/Mdns.h>
+
+namespace chip {
+namespace Mdns {
+
+constexpr uint64_t kUndefinedNodeId = 0;
+
+class ServicePool
+{
+public:
+    class Entry
+    {
+    public:
+        friend ServicePool;
+
+        Entry();
+        Entry(Entry && rhs);
+        Entry & operator=(Entry && rhs);
+        MdnsService mService;
+
+        uint64_t GetNodeId() { return mNodeId; }
+        uint64_t GetFabricId() { return mFabricId; }
+
+        ~Entry();
+
+    private:
+        void Clear();
+
+        Entry & MoveFrom(Entry && rhs);
+        CHIP_ERROR Emplace(const MdnsService & service, uint64_t nodeId, uint64_t fabricId);
+        CHIP_ERROR Emplace(MdnsService && service, uint64_t nodeId, uint64_t fabricId);
+
+        bool mHasValue;
+        bool mHasPending;
```

Comment:
`mHasPending` ... what?  This should be called `mHasPendingDeletion` or `mIsMarkedDeleted` or `mIsTombstone` if we're OK assuming people know what that means in the context of hashtables or something like that, imo.

Reply:
`mIsTombstone` is a good name.